### PR TITLE
insighttoolkit: migrate from homebrew-science

### PR DESCRIPTION
**Pull request by iMichka**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
It's version 4.12.2, science had version 4.12.1.

I already cleaned up a lot of stuff.

The main changes are:
- not fftw, because that option was not used (analytics say so), and fftw is GPL, which makes everything built on top of itk GPL too
- Python build by default. Build time is quite long, let's see how long on our bot
- c++11 by default as vtk is c++11 only

**Comment by iMichka** [comment]
Companion PR: https://github.com/Homebrew/homebrew-science/pull/6322

**Comment by apjanke** [comment]
The test builds are failing; this was in the output.

```
04:07:47 Error: No available formula with the name "gdcm" 
04:07:47 Build step 'Execute shell' marked build as failure
```

I believe that's because `gdcm` is still in `homebrew/science`:
```
$ brew info gdcm
homebrew/science/gdcm: stable 2.6.8 (bottled)
```

and we don't support dependencies from homebrew-core to other taps. At the least, the formula name would have to be qualified in the dependency declaration.

To get this to work, I believe you'd need to migrate `gdcm` over to core as well.

File: Formula/insighttoolkit.rb [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  depends_on :python => :recommended if MacOS.version <= :snow_leopard
+  depends_on "hdf5"
+  depends_on "jpeg"
+  depends_on "libpng"
+  depends_on "libtiff"
+  depends_on "gdcm"
+  depends_on "opencv"
+
+  if build.with? "python3"
+    depends_on "vtk" => ["with-python3", "without-python"]
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: will fail audit.

File: Formula/insighttoolkit.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  elsif build.with? "python"
+    depends_on "vtk" => ["with-python"]
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: will fail audit.

File: Formula/insighttoolkit.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
+class Insighttoolkit < Formula
+  desc "ITK is a toolkit for performing registration and segmentation"
+  homepage "https://www.itk.org"
+  url "https://downloads.sourceforge.net/project/itk/itk/4.12/InsightToolkit-4.12.2.tar.gz"
+  sha256 "8ba2183111723c4405af80724c60d07484c2bff8120ee4cef543bb726d07855e"
+  head "git://itk.org/ITK.git"
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: we'll need to do without a head spec if there's no https

File: Formula/insighttoolkit.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
+class Insighttoolkit < Formula
+  desc "ITK is a toolkit for performing registration and segmentation"
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: don't need "ITK is a"

File: Formula/insighttoolkit.rb [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  homepage "https://www.itk.org"
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: should have trailing slash

File: Formula/insighttoolkit.rb [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  option "without-python", "Build without python2 support"
+
+  depends_on "cmake" => :build
+  depends_on :python3 => :optional
+  depends_on :python => :recommended if MacOS.version <= :snow_leopard
+  depends_on "hdf5"
+  depends_on "jpeg"
+  depends_on "libpng"
+  depends_on "libtiff"
+  depends_on "gdcm"
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: non-build deps should be alphabetized amongst themselves

File: Formula/insighttoolkit.rb [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    # Build step
+    system ENV.cxx, "-isystem", "#{include}/ITK-#{v}", "-o", "test.cxx.o", "-c", "test.cxx"
+    # Linking step
+    system ENV.cxx, "test.cxx.o", "-o", "test",
+                    "#{lib}/libITKCommon-#{v}.1.dylib",
+                    "#{lib}/libITKVNLInstantiation-#{v}.1.dylib",
+                    "#{lib}/libitkvnl_algo-#{v}.1.dylib",
+                    "#{lib}/libitkvnl-#{v}.1.dylib"
+    system "./test"
+    system "python", "-c", "import itk;itk.Image[itk.UC, 2]()"
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: python3 should be tested too

File: Formula/insighttoolkit.rb [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        image->Update();
+
+        return EXIT_SUCCESS;
+      }
+    EOS
+
+    v=version.to_s.split(".")[0..1].join(".")
+    # Build step
+    system ENV.cxx, "-isystem", "#{include}/ITK-#{v}", "-o", "test.cxx.o", "-c", "test.cxx"
+    # Linking step
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: this seems weird. why can't it be in one step

File: Formula/insighttoolkit.rb [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+      -DModule_ITKVtkGlue=ON
+      -DITK_USE_GPU=ON
+    ]
+
+    mkdir "build" do
+      if build.with?("python") || build.with?("python3")
+        python_executable = `which python`.strip if build.with? "python"
+        python_executable = `which python3`.strip if build.with? "python3"
+
+        python_prefix = `#{python_executable} -c 'import sys;print(sys.prefix)'`.chomp
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: this is messy. we shouldn't need to resort to these kind of hacks.

File: Formula/insighttoolkit.rb [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    depends_on "vtk" => ["with-python"]
+  else
+    depends_on "vtk"
+  end
+
+  needs :cxx11
+
+  def install
+    args = std_cmake_args + %W[
+      ..
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: this should be outside the args array

File: Formula/insighttoolkit.rb [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+        python_prefix = `#{python_executable} -c 'import sys;print(sys.prefix)'`.chomp
+        python_include = `#{python_executable} -c 'from distutils import sysconfig;print(sysconfig.get_python_inc(True))'`.chomp
+        python_version = "python" + `#{python_executable} -c 'import sys;print(sys.version[:3])'`.chomp
+
+        args << "-DITK_WRAP_PYTHON=ON"
+        args << "-DPYTHON_EXECUTABLE='#{python_executable}'"
+        args << "-DPYTHON_INCLUDE_DIR='#{python_include}'"
+        # CMake picks up the system's python dylib, even if we have a brewed one.
+        if File.exist? "#{python_prefix}/Python"
```
Comment (ilovezfs) [inline Formula/insighttoolkit.rb]: should use Pathname's exist?

**Comment by iMichka** [comment]
Thanks for the review. I'll go through the different points later today

**Comment by iMichka** [comment]
I opened a pull request for `gdcm`: https://github.com/Homebrew/homebrew-core/pull/18352
I'll update `insightoolkit` once `gdcm` is in.

File: Formula/insighttoolkit.rb (hunk 1 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: Not sure what I can do about it. `vtk` is either/or, and `insighttoolkit` the same. I know it because I ported `insighttoolkit` to python 3 back then. It was never ment to be able to build both at the same time.

I would personally prefer to go full Python 3 everywhere.

File: Formula/insighttoolkit.rb (hunk 3 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: Will update to https://itk.org/ITK.git

File: Formula/insighttoolkit.rb (hunk 4 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: Done

File: Formula/insighttoolkit.rb (hunk 6 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: Done

File: Formula/insighttoolkit.rb (hunk 9 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: We did the same for vtk. I do not like it neither. I always hoped homebrew would propose something in the DSL to hide that boilerplate code.

We do support all pythons in the path, and need to tell cmake where python is. So if you have anaconda Python, or system Python, or upstream Python ... you have to rely on some hacks like these ...

File: Formula/insighttoolkit.rb (hunk 8 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: Let me try again but I had the feeling that it was the only way to get this to build ...

File: Formula/insighttoolkit.rb (hunk 10 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: The `..`? Done.

File: Formula/insighttoolkit.rb (hunk 11 above)
Comment (iMichka) [reply Formula/insighttoolkit.rb]: We use the same syntax in `vtk` and `root`.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
Build failed.
https://gist.githubusercontent.com/ilovezfs/dc4bd62610311154f27012acb6f10888/raw/e8af25d3f4d081e6781caa3c47b00801ce5412d0/gistfile1.txt
https://gist.githubusercontent.com/ilovezfs/d1790afd779c144a048097c457325376/raw/50ad6afa23238aa9dcd6e021896bffd59b2c890e/gistfile1.txt
https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/10403/version=sierra/consoleFull

**Comment by ilovezfs** [comment]
Also
```
insighttoolkit:
  * Dependency vtk does not define option "with-python"
  * Dependency vtk does not define option "with-python"
  * Dependency vtk should not use option with-python3
  * Dependency vtk should not use option without-python
  * Dependency vtk should not use option with-python
  * Use `assert_predicate <path_to_file>, :exist?` instead of `File.exist?`
  * Use `assert_predicate <path_to_file>, :exist?` instead of `File.exist?`
  * Use `assert_predicate <path_to_file>, :exist?` instead of `File.exist?`
```

**Comment by ilovezfs** [comment]
looks like it built on 10.11 so it's probably an Xcode 9 issue
